File: 2026/January/Day-03-Flattening-a-Linked-List/cpp/solution.cpp
class Solution {
  public:
  
    // Helper function to merge two sorted bottom-linked lists
    Node* merge(Node* a, Node* b) {
        if (a == NULL) return b;
        if (b == NULL) return a;

        Node* result;

        if (a->data < b->data) {
            result = a;
            result->bottom = merge(a->bottom, b);
        } else {
            result = b;
            result->bottom = merge(a, b->bottom);
        }

        // Important: next pointer should be NULL
        result->next = NULL;
        return result;
    }

    Node* flatten(Node* root) {
        // Base case
        if (root == NULL || root->next == NULL)
            return root;

        // Flatten the list on the right
        root->next = flatten(root->next);

        // Merge current list with flattened right list
        root = merge(root, root->next);

        return root;
    }
};

File: 2026/January/Day-03-Flattening-a-Linked-List/python/solution.py
class Solution:
    
    # Merge two sorted bottom-linked lists
    def merge(self, a, b):
        if not a:
            return b
        if not b:
            return a

        if a.data < b.data:
            result = a
            result.bottom = self.merge(a.bottom, b)
        else:
            result = b
            result.bottom = self.merge(a, b.bottom)

        result.next = None  # important
        return result

    def flatten(self, root):
        # Base case
        if not root or not root.next:
            return root

        # Flatten the right list
        root.next = self.flatten(root.next)

        # Merge current list with flattened list
        root = self.merge(root, root.next)

        return root
